sycl/source/detail/scheduler/commands.cpp

```diff
@@ -681,15 +681,12 @@ static void adjustNDRangePerKernel(NDRDescT &NDR, RT::PiKernel Kernel,
 
   if (WGSize[0] == 0) {
     // kernel does not request specific workgroup shape - set one
-    // TODO maximum work group size as the local size might not be the best
-    //      choice for CPU or FPGA devices
-    size_t WGSize1D = get_kernel_work_group_info<
-        size_t,
-        cl::sycl::info::kernel_work_group::work_group_size>::get(Kernel,
-                                                                 Device);
-    assert(WGSize1D != 0);
+    id<3> MaxWGSizes = get_device_info<
+        id<3>, cl::sycl::info::device::max_work_item_sizes>::get(Device);
```

Comment:
Please be careful here. If I understand correctly, `max_work_item_sizes` is kernel-agnostic query: particular kernel might require additional resources and cannot be called with ND-range obtained by this query (from OpenCL 1.2, clEnqueueNDRange):

> CL_OUT_OF_RESOURCES if there is a failure to queue the execution instance of kernel
on the command-queue because of insufficient resources needed to execute the kernel.
For example, the explicitly specified local_work_size causes a failure to execute the
kernel because of insufficient resources such as registers or local memory.

`work_group_info` returns a value which accounts kernel complexity to calculate work-group size: from OpenCL 1.2, description of `CL_KERNEL_WORK_GROUP_SIZE`:

> This provides a mechanism for the application to query the maximum work-group size that can be used to execute a kernel on a specific device given by device. The OpenCL implementation uses the resource requirements of the kernel(register usage etc.) to determine what this work-group size should be.

I guess the problem is that spelling is not quite good: what does "maximum work-group size" mean: is it max size across one particular dimension, or across all, or it is just a total number of work-items withing a work-group? We probably need a spec lawyer here (@bashbaug?)

Reply:
The problem here is if `WGSize1D > MaxWGSizes[0]` we would get a failure. You got a point that if `MaxWGSizes[2] < WGSize1D`  we would likewise have a problem. A solution could be to do something along the lines of
```cpp
size_t WGSizeX = std::min(WGSize1D, MaxWGSizes[0]);
size_t WGSizeY = std::min(WGSize1D / WGSizeX, MaxWGSizes[1]);
size_t WGSizeZ = WGSize1D / (WGSizeX * WGSizeY);

assert(WGSizeZ <= MaxWGSizes[2]);

WGSize = {WGSizeX, WGSizeY, WGSizeZ};
```
but this assumes that `WGSize1D` is divisible by `WGSizeX` and by `WGSizeX * WGSizeY`. If we cannot assume this we will have to be more creative.

Reply:
i don't think this code can choose the best work-groups size of a kernel, the task here is to chose legal and not the worst size. 
So, can we make the following:
WGSize = {min(WGSize1D,  MaxWGSizes[2], 1, 1) ?

Reply:
If "good enough" is good enough then I'm all for it. 👍

Well remembered that we recently had the `MaxWGSizes` reversed (https://github.com/intel/llvm/pull/1177)! However, I think this code actually assume that to be the case before as well, but didn't fail because for most CUDA devices it was using the smallest max WG size.

Reply:
I suggest then to go with @romanovvlad's proposal

Reply:
We are reversing sizes after this call is done, on line 936.
@steffenlarsen  Are you ok with the suggested solution or you have some concerns?

Reply:
> We are reversing sizes after this call is done, on line 936.

But the max work item sizes are also reversed by the PI query, so the actual max at index 2 in `MaxWGSizes[0]`, which if we use that for `WGSize[0]` it will be moved back to index 2, where it is the valid max.

Other than that I have no problems with this approach, but I don't have access to change this PR so I'll ping @StuartDAdams to have this put in. Thank you for your feedback!

Reply:
You are right. So, we should do:
WGSize = {1, 1, min(WGSize1D, MaxWGSizes[2])
and write a comment why we are doing so.

Reply:
[Done!](https://github.com/intel/llvm/pull/952/commits/9897d549eb37db79c70b41c3d82a61971f06a608)